stetho/src/main/java/com/facebook/stetho/inspector/network/NetworkEventReporterImpl.java

```diff
@@ -115,15 +128,29 @@ private static String readBodyAsString(
       InspectorRequest request) {
     try {
       byte[] body = request.body();
-      if (body != null) {
-        return new String(body, Utf8Charset.INSTANCE);
+      if (body == null || body.length == 0) {
+        return "";
+      }
+      try {
+        CharBuffer charBuffer = decoder.decode(ByteBuffer.wrap(body));
+        return charBuffer.toString();
+      } catch (CharacterCodingException e) {
+        String logMessage = "Charset in POST/PUT is not UTF-8. Data (length:"+ body.length
+                +") cannot be represented as a string. ";
+        CLog.writeToConsole(
+                peerManager,
+                Console.MessageLevel.WARNING,
+                Console.MessageSource.NETWORK,
+                 logMessage + e);
+        return logMessage;
```

Comment:
Hmm, I like returning the log message here, but it suggests actually that we'd be better off merging the error cases below:

````
try {
 // decode...
} catch (IOException | OutOfMemoryError e) {
  String error = determineErrorMessage(e);
  CLog.writeToConsole("Could not reproduce POST/PUT body: " + error);
  return error;
}
````

Then determineErrorMessage could peak and see if it's a CharsetCodingException and craft your better error message above.